test/e2e/main_test.go

```diff
@@ -72,6 +85,71 @@ func TestMain(m *testing.M) {
 		// tests that need the tracing in place.
 		defer zipkin.CleanupZipkinTracingSetup(log.Printf)
 
-		return m.Run()
+		if code := m.Run(); code > 0 {
+			return code
+		}
+
+		// The knative/pkg base controller emits custom metrics, plus some other components can
+		// potentially emit custom metrics. By default custom metrics should not be published to
+		// Stackdriver.
+		// After running all tests, we verify that no custom metrics have been emitted in the past
+		// 30 min to cover roughly how long the e2e tests run.
+		return verifyNoCustomMetrics(projectID, time.Duration(-30)*time.Minute)
 	}())
 }
+
+// Return code meaning
+// 0: success, no custom metrics found
+// 1: failure, found custom metrics
+// 2: failure, unable to verify custom metrics
+func verifyNoCustomMetrics(projectID string, duration time.Duration) int {
```

Comment:
maybe just return an error instead of a code?

Reply:
The reason the code is returned here is that there are multiple return values. I could return multiple errors, but then the TestMain has to translate errors to codes. 

But if you think that's better, I am happy to do that.

Reply:
Can main just call log.Fatal when there is an error?

Reply:
Changed to error.

The previous logic differentiates different return codes, however that's not very important. I think log.Fatal with error message works too.